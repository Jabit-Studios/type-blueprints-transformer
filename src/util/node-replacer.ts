import ts from "typescript";

/**
 * Predicate accepts ts.Node or it"s subtypes
 */
type NodePredicate<T extends ts.Node = ts.Node> = (node: T) => boolean

/**
 * Returns mapping function that replaces
 * nodes that satisfy the passed predicate
 * with a replacement node passed but skips the first appearance
 */
export function getNodeReplacer(predicate: NodePredicate, replacement: ts.Node) {
    let replaced = false;
    return (node: ts.Node) => {
        if (replaced && predicate(node)) return [replacement, replaced]
        if (replaced || !predicate(node)) return [node, replaced]

        replaced = true;
        return [replacement, replaced]
    }
}