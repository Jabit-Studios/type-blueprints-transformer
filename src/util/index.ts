import { getNodeReplacer } from "./node-replacer";

export { getNodeReplacer };