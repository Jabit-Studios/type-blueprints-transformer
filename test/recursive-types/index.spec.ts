import { readFileSync } from "fs";
import { it, describe, expect } from "vitest";
import { compile } from "../util";

describe("Recursive Types", () => {
	it("should allow for one level deep recursion", () => {
		const output = compile("test/recursive-types/snapshots/single-deep.ts");
		expect(output).toMatchSnapshot();
	});
});
