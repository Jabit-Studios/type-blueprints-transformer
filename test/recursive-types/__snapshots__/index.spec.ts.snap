// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Recursive Types > should allow for one level deep recursion 1`] = `
";
import { $terrify } from "./test";
interface SingleDeep {
    a: SingleDeep;
}
const singleDeep = $stamp<SingleDeep>();
"
`;
